### ISSUE TITLE
A11y_.NETCore3.0(WPF)_WPFBrush_Contents_Keyboard: The treeview items with toggle button receive keyboard focus twice while navigating using the arrow key.

### ISSUE THREAD

**Issue by grubioe** [dotnet/wpf#1455]
Copied from VSO [948205](https://devdiv.visualstudio.com/DevDiv/_queries/edit/948205/)

Check out Accessibility Insights! - Identify accessibility bugs before check-in and make bug fixing faster and easier.”

Environment Details:
Application Name: .NET Core 3.0 (WPF and ASP.NET Core)

Application version: VS 2019 Preview 4.7.03190

Windows Version: Windows10

ScreenReader Version: Narrator, NVDA 2019.1.1

Pre-requisite:

1) Download the latest .NET Core 3 daily build by going to this link.
2) Install Windows (x86 or x64) versions of sdk listed under Master (3.0.x Runtime) by clicking "Installer" link.

Repro Steps:

1) Open Visual Studio.
2) Hit on File and select Open->Project/Solution.
3) Select the "Brushes" solution and open it. 
4) From the Solution Explorer, right click on the file under the solution "Brushes".
5) Select "Debug" and hit "Start new Instance".
6) Main Window for "Brush Examples" solution will appear.
7) Navigate to "Contents" section and Expand it.
8) Navigate to each of the treeview items with toggle button using the arrow keys.
9) Observe that these element receive keyboard focus twice.

Actual:

When keyboard users navigate through the treeview items with toggle buttons of the "Contents" using the arrow keys, these elements receive keyboard focus twice. On first keyboard focus (blue boarder) the user is not able to activate the toggle button and on second keyboard focus (dotted line border) the user is able to activate the toggle button.

Expected:

The treeview items with toggle button should receive keyboard focus once and the user should be able to activate the toggle button.

User Impact:

If an interactive element receive keyboard focus twice and it is actionable for keyboard users on one of the keyboard focus then the users will get confused and might not know how to interact with them.

Recommendations:

Please refer to below URL which is repository for accessibility bug fixes.

https://microsoft.sharepoint.com/teams/msenable/Pages/Home.aspx

MAS Reference:

MAS 2.1.1

https://microsoft.sharepoint.com/:w:/r/teams/msenable/_layouts/15/WopiFrame.aspx?sourcedoc={8492c4eb-c179-40ae-8777-cd044ed725a2}

**Comment by rladuca** [comment]
While this is an application bug with the control template for the TreeViewItem, there is a similar problem with the default template for TreeViewItem.

When navigating via Narrator, you will hit both the Header item and the ToggleButton used as the expander.

When navigating via NVDA this is slightly different.  NVDA will move directly to the next TreeViewItem and will not select the internal buttons.  This is generally consistent with arrow key navigation in TreeView.

**Comment by rladuca** [comment]
In terms of activation, Enter is generally not used to expand a TreeViewItem.  This is done via the arrow keys.  We have to figure out what that will do with regards to this.  We could handle Enter in TreeViewItem with a compat flag, I don't believe this will negatively impact anyone and should allow Narrator to activate TreeViewItems.

**Comment by rladuca** [comment]
Given that NVDA navigates appropriately (as in, it moves from TreeViewItem to TreeViewItem and does not stop on the internal ToggleButton) we won't be fixing the first issue specified here.

However, neither Narrator nor NVDA is able to activate a TreeViewItem.  This is because TreeViewItem does not respond to the Return key.  This can be added (albeit with a compat flag) reasonably easily.  We should fix that.

**Comment by weltkante** [comment]
> We could handle Enter in TreeViewItem with a compat flag, I don't believe this will negatively impact anyone and should allow Narrator to activate TreeViewItems.

This is not how explorer works (enter does not expand, it "commits" the selection and activates the corresponding view), and I'm sure there are people already handling enter keys similarly.

Why can't WPF handle accessibility in tree views like everyone else does, why does it have to add a fake compat flag and key handling when nobody else does this? I mean its ok if that compat flag is a short term solution and disabled by default, but if its disabled by default would it really help with accessibility?

**Comment by rladuca** [comment]
@weltkante I revisted this just thinking about the activation and Narrator seems to be able to activate the items just fine using Expand/Collapse via UIA.  So we don't need a change in that area.  I probably mixed up several samples I was working with at the time, so that's my fault and your comment made me reconsider my conclusions (so thanks!).

The general story for compat flags (at this point) is that we would hide accessibility changes behind these flags for 3.1, much as we did for 4.8 (the UseLegacyAccessibility.X flags).  The reason for that in 3.0/3.1 is to provide as close to full compat with .NET 4.8 as we reasonably can to help support porting from .NET Framework.  These flags would then turn to default on in .NET 5.0, where the compat story becomes less about .NET 4.8 and developers have the option to publish self-contained on 3.1 if they want to insulate from changes.

Overall this bug can be closed since Narrator can activate the base TreeView items with its activation key.  Also, Narrator itself is making a mistake to focus on the ToggleButton as its not marked focusable so it should not do so.
